packages/adblocker/src/lists.ts

```diff
@@ -136,13 +138,24 @@ export function f(strings: TemplateStringsArray): NetworkFilter | CosmeticFilter
 export function parseFilters(
   list: string,
   config: Partial<Config> = new Config(),
-): { networkFilters: NetworkFilter[]; cosmeticFilters: CosmeticFilter[] } {
+): {
+  networkFilters: NetworkFilter[];
+  cosmeticFilters: CosmeticFilter[];
+  preprocessors: Preprocessor[];
+} {
   config = new Config(config);
 
   const networkFilters: NetworkFilter[] = [];
   const cosmeticFilters: CosmeticFilter[] = [];
   const lines = list.split('\n');
 
+  const preprocessors: Preprocessor[] = [];
+  // The filters affected by a preprocessor
+  const filterStack: IFilter[] = [];
+  // This defines the start index in `filtersInBlock` of each preprocessor block
+  // The highest level of block will have an index of zero
+  const preprocessorStack: { start: number; condition: string }[] = [];
+
```

Comment:
We need to collect preprocessors which are already created. So there need to be two preprocessors array at least. For the 'preprocessorStack', they're meant to express the current depth of preprocessor block.

Reply:
I think it could be simpler to replace:

```typescript
const filterStack: IFilter[] = [];
const preprocessorStack: { start: number; condition: string }[] = [];
```

With a single:
```typescript
const preprocessorStack: Preprocessor[] | undefined = undefined;
```

* When you encounter a new `BEGIF`, you start a new `preprocessorStack` with an empty preprocessor.
* When you encounter a network/cosmetic filter, you push its `filterID` to the last element of `preprocessorStack`:
```typescript
if (preprocessorStack !== undefined && preprocessorStack.length !== 0) {
  preprocessorStack[preprocessorStack.length - 1]. filterIDs.add(filter.getId());
```
* When you find nested `BEGIF` you push a new `Preprocessor`:
```typescript
preprocessorStack.push(new Preprocessor({
condition: `${conditionOfParent} && {conditionOfNewPreprocessor}`
}))
```
* When you encounter `ELSE` or `ENDIF` you pop one element from `preprocessorStack` and add it to `preprocessors`:
```typescript
preprocessors.push(preprocessorStack.pop());
```

I think this will simplify the state you need to keep but also the logic you have below when you encounter `ELSE` or `ENDIF`.